File: python/leetcode/354.py
class Solution(object):
    def maxEnvelopes(self, envelopes):
        def bin_search(A, key):
            l, r = 0, len(A)
            while l < r:
                mid = (l + r) / 2
                if A[mid][1] < key[1]: l = mid + 1
                else: r = mid
            return l

        envelopes.sort(
            cmp=lambda x, y: x[0] - y[0] if x[0] != y[0] else y[1] - x[1])
        n = len(envelopes)
        tails = []
        for i in range(n):
            e = envelopes[i]
            p = bin_search(tails, e)
            if p == len(tails): tails.append(e)
            else: tails[p] = e
        return len(tails)

File: cpp/leetcode/354.cpp
class Solution {
public:
    static bool cmp_first(const pair<int, int>& i, const pair<int, int>& j) {
        if (i.first == j.first) return i.second > j.second;
        return i.first < j.first;
    }
    int maxEnvelopes(vector<pair<int, int>>& envelopes) {
        vector<int> candidates;
        sort(envelopes.begin(), envelopes.end(), cmp_first);
        vector<int> dp;
        for (int i = 0; i < envelopes.size(); ++i) {
            auto itr = lower_bound(dp.begin(), dp.end(), envelopes[i].second);
            if (itr == dp.end()) dp.push_back(envelopes[i].second); 
            else *itr = envelopes[i].second;
        }
        return dp.size();
    }
};
